include/swift/Sema/IDETypeChecking.h

```diff
@@ -206,6 +206,36 @@ namespace swift {
   /// the decl context.
   ProtocolDecl *resolveProtocolName(DeclContext *dc, StringRef Name);
 
+  /// Reported type of a variable declaration.
+  struct VariableTypeInfo {
+    /// The start of the variable identifier.
+    uint32_t Offset;
+
+    /// The length of the variable identifier.
+    uint32_t Length;
+
+    /// Whether the variable has an explicit type annotation.
+    bool HasExplicitType;
+
+    /// The start of the printed type in a separate string buffer.
+    uint32_t TypeOffset;
+
+    /// The length of the printed type in a separate string buffer.
+    uint32_t TypeLength;
+
+    VariableTypeInfo(uint32_t Offset, uint32_t Length, bool HasExplicitType,
+                     uint32_t TypeOffset, uint32_t TypeLength);
+  };
+
+  /// Collect type information for every variable declaration in \c SF
+  /// within the given range, defined by \c Offset and \c Length.
+  /// All types will be printed to \c OS and the type offsets of the
+  /// \c VariableTypeInfos will index into the string that backs this
+  /// stream.
+  ArrayRef<VariableTypeInfo> collectVariableType(
+      SourceFile &SF, Optional<unsigned> Offset, Optional<unsigned> Length,
```

Comment:
Just curious, is there a reason to use `SourceRange` over `CharSourceRange` in this case, which is already used in `walkToDeclPre`?

Reply:
I chose `SourceRange` because all AST nodes already store their `SourceRange` but not their `CharSourceRange`. If you want to convert `SourceRange` to `CharSourceRange`, you need to reach out to the lexer again to determine the length of the end token. 

Thus, I think it’s better to just do the conversion once for the requested range and avoid converting the range of every AST node.